### ISSUE TITLE
`BerSequence` and `DerSequence` derive macros produce code that can panic

### ISSUE THREAD

**Issue by plusvic** [rusticata/asn1-rs#40]
The code generated by both [BerSequence](https://docs.rs/asn1-rs/0.6.1/asn1_rs/derive.BerSequence.html) and [DerSequence](https://docs.rs/asn1-rs/0.6.1/asn1_rs/derive.DerSequence.html) for the `TryFrom<Any>` trait uses the [finish](https://docs.rs/nom/latest/nom/trait.Finish.html) function for error convertion.  The issue with `finish` is that it can panic when the parser's result is `Err(Err::Incomplete(_))`. This is problematic when parsing ASN.1 structures that are corrupted. 

I've found this issue while trying to use the [x509-parser](https://github.com/rusticata/x509-parser) crate for parsing X509 certificates in Windows PE files. The files are sometimes corrupt, and make the program panic. 

Consider the [AlgorithmIdentifier](https://docs.rs/x509-parser/0.16.0/x509_parser/x509/struct.AlgorithmIdentifier.html#) structure:
```rust
#[derive(Clone, Debug, PartialEq, DerSequence)]
#[error(X509Error)]
pub struct AlgorithmIdentifier<'a> {
    #[map_err(|_| X509Error::InvalidAlgorithmIdentifier)]
    pub algorithm: Oid<'a>,
    #[optional]
    pub parameters: Option<Any<'a>>,
}
```

The code generated by `DerSequence` for the `try_from` function looks like:

```rust
fn try_from(any: Any<'ber>) -> asn1_rs::Result<Self, X509Error> {
    use asn1_rs::nom::*;
    any.tag().assert_eq(Self::TAG)?;
    let i = any.data;
    let (i, algorithm) = FromBer::from_ber(i).finish().map_err(|_| X509Error::InvalidAlgorithmIdentifier)?;
    let (i, parameters) = FromBer::from_ber(i).finish()?;
    let _ = i;
    Ok(Self {
        algorithm,
        parameters,
    })
}
```

Notice the use of `finish` after parsing both the `algorithm` and `parameters` fields. The documentation for `finish` says:

> warning: if the result is Err(Err::Incomplete(_)), this method will panic.
> 
> “complete” parsers: It will not be an issue, Incomplete is never used
> “streaming” parsers: Incomplete will be returned if there’s not enough data for the parser to decide, and you should gather more data before parsing again. Once the parser returns either Ok(_), Err(Err::Error(_)) or Err(Err::Failure(_)), you can get out of the parsing loop and call finish() on the parser’s result

So, it looks like "streaming" parsing is being used, but I'm not sure if that's the case, or why.

**Comment by chifflier** [comment]
This is indeed a serious problem (unwanted `panic` in a library), thanks for reporting it
I am investigating for a solution

**Comment by plusvic** [comment]
Thank you for the fix! Any plans to publish a release that includes this fix?

**Comment by chifflier** [comment]
The fix has been published in `asn1-rs-derive` 0.5.1
A simple `cargo update` on this crate should be enough (there is no breaking change).

**Comment by plusvic** [comment]
Great! Thank you!
